### PR TITLE
Speedup sympy.physics.vector

### PR DESCRIPTION
**Pull request by isuruf**
I get about a 1.25x speedup (with SymEngine) for `n_link_pendulum_on_cart`. Speedup with SymPy core is 1.05x.

### CODE REVIEW

**Comment by isuruf** [comment]
I've updated the speedups. It's less than I mentioned because there was an change in how I compiled symengine between the 2 runs.
@sympy/mechanics what do you think about converting the internal storage to a dictionary?

**Review by moorepants — APPROVED** [review]
+1 if tests pass.

**Comment by isuruf** [comment]
Doctests fail because the order of the args is different now.

**Comment by isuruf** [comment]
Since I'm using dictionaries, order of `args` differs between multiple runs.  How should I handle the doctest failures? Make printing independent of the order of args or some other way?

**Comment by jksuom** [comment]
Perhaps you could write `for k, v in ordered(d.items()):`.

**Comment by isuruf** [comment]
@jksuom, do you mean when `self.args` is made? It's a waste to just order them when the only usefulness of it is for printing.

**Comment by jksuom** [comment]
Yes, that was my first thought. But if the issue is only in (string) printing, then you could order the arguments in `__str__`.

**Comment by isuruf** [comment]
How do I order the arguments? `self.args` is a list of `(ImmutableMatrix, ReferenceFrame)` pairs

**Comment by jksuom** [comment]
I believe that `ordered` could be made to work on such pairs. (Though I actually don't know its internals.)

**Comment by isuruf** [comment]
`ReferenceFrame` is hard to work with because it has a cyclic reference to itself. Easiest way to order `ReferenceFrame`s were to introduce an index for the `ReferenceFrame`s and that's what I did here.

**Comment by isuruf** [comment]
Test failure is due to an unrelated bug, #11684. Shall I rerun the test on travis, because this bug can be reproduced with the correct PYTHONHASHSEED and random seed?

**Comment by isuruf** [comment]
@moorepants, ping. This would affect dostests in PyDy as well.

File: doc/src/modules/physics/mechanics/examples/bicycle_example.rst [hunk 1]
```diff
@@ -115,7 +115,7 @@ again are calculated automatically when first needed. ::
...
...q2)*cos(q5) + sin(q5)*cos(htangle + q4)*cos(q2)) + 1)*Y.y + WRrad*(u1*sin(q2) + u3 + u4)*R.x - WRrad*u2*R.y + framelength*(u1*sin(q2) + u4)*Frame.x - framelength*(-u1*sin(htangle + q4)*cos(q2) + u2*cos(htangle + q4))*Frame.y + (-WFrad*(sin(q2)*cos(q5) + sin(q5)*cos(htangle + q4)*cos(q2))*((-sin(q2)*sin(q5) + cos(htangle + q4)*cos(q2)*cos(q5))*u1 + u2*sin(htangle + q4)*cos(q5) - u4*sin(q5))/sqrt((-sin(q2)*cos(q5) - sin(q5)*cos(htangle + q4)*cos(q2))*(sin(q2)*cos(q5) + sin(q5)*cos(htangle + q4)*cos(q2)) + 1) + forkoffset*((sin(q2)*cos(q5) + sin(q5)*cos(htangle + q4)*cos(q2))*u1 + u2*sin(htangle + q4)*sin(q5) + u4*cos(q5)))*Fork.x + (forklength*((-sin(q2)*sin(q5) + cos(htangle + q4)*cos(q2)*cos(q5))*u1 + u2*sin(htangle + q4)*cos(q5) - u4*sin(q5)) - forkoffset*(-u1*sin(htangle + q4)*cos(q2) + u2*cos(htangle + q4) + u5))*Fork.y + (WFrad*(sin(q2)*cos(q5) + sin(q5)*cos(htangle + q4)*cos(q2))*(-u1*sin(htangle + q4)*cos(q2) + u2*cos(htangle + q4) + u5)/sqrt((-sin(q2)*cos(q5) - sin(q5)*cos(htangle + q4)*cos(q2))*(sin(q2)*cos(q5) + sin(q5)*cos(htangle + q4)*cos(q2)) + 1) - forklength*((sin(q2)*cos(q5) + sin(q5)*cos(htangle + q4)*cos(q2))*u1 + u2*sin(htangle + q4)*sin(q5) + u4*cos(q5)))*Fork.z
```
Comment (moorepants) [inline doc/src/modules/physics/mechanics/examples/bicycle_example.rst]: Do we know this is equivalent?

File: doc/src/modules/physics/vector/kinematics.rst [hunk 2]
```diff
@@ -468,7 +468,7 @@ other. This is coded as: ::
   >>> C.set_ang_vel(B, -u2 * B.z)
   >>> D.set_ang_vel(C, u3 * C.y)
   >>> D.ang_vel_in(N)
-  u3*C.y - u2*B.z + u1*A.x
+  u1*A.x - u2*B.z + u3*C.y
```
Comment (moorepants) [inline doc/src/modules/physics/vector/kinematics.rst]: This is fine. There is no requirement on order for Vector components.

File: sympy/physics/vector/frame.py [hunk 3]
```diff
@@ -198,6 +199,8 @@ def __init__(self, name, indices=None, latexs=None, variables=None):
         self.varlist = (CoordinateSym(variables[0], self, 0), \
                         CoordinateSym(variables[1], self, 1), \
                         CoordinateSym(variables[2], self, 2))
+        ReferenceFrame._count += 1
+        self.index = ReferenceFrame._count
```
Comment (moorepants) [inline sympy/physics/vector/frame.py]: What is the purpose of this change?

File: sympy/physics/vector/vector.py [hunk 4]
```diff
@@ -345,12 +339,20 @@ def __ror__(self, other):
     def __rsub__(self, other):
         return (-1 * self) + other
 
-    def __str__(self, printer=None):
+    def __str__(self, printer=None, order=True):
```
Comment (moorepants) [inline sympy/physics/vector/vector.py]: It seems that the order is respect to this new index for the reference frame. Why would a user want an index for the frames and want them to be ordered in that way when printing?

**Review by moorepants — COMMENTED** [review]
Left some comments.

File: doc/src/modules/physics/mechanics/examples/bicycle_example.rst (hunk 1 above)
Comment (isuruf) [inline doc/src/modules/physics/mechanics/examples/bicycle_example.rst]: Yes, I checked by subtracting the two expressions. It'll be great if somebody could do it as well to be sure.

File: sympy/physics/vector/vector.py (hunk 4 above)
Comment (isuruf) [inline sympy/physics/vector/vector.py]: I made this change because the dictionary is unordered and each runs prints the same vector differently. Index was to make sure that the printing is the same across multiple runs. Let me know if there is a better way.

Comment (isuruf) [inline sympy/physics/vector/vector.py]: @moorepants, ping

File: sympy/physics/vector/frame.py (hunk 3 above)
Comment (certik) [inline sympy/physics/vector/frame.py]: @isuruf can you reply here?

File: sympy/physics/vector/vector.py (hunk 4 above)
Comment (certik) [inline sympy/physics/vector/vector.py]: @isuruf I thought SymPy gets around this by ordering the dictionaries before printing, if you use a Printer.

**Review by certik — APPROVED** [review]
It looks good to me.

File: sympy/physics/vector/frame.py (hunk 3 above)
Comment (isuruf) [inline sympy/physics/vector/frame.py]: Same as below. To order the printing
